wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+        ret = iotsafe_verify_hash(pubkey_slot,
+                IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+                hash, hashSz, sig_raw, 2 * IOTSAFE_ECC_KSIZE,
+                result);
+    } else {
+        word32    idx = 0;
+        ecc_key   myKey;
+        ret = wc_ecc_init(&myKey);
+        if (ret == 0) {
+            ret = wc_EccPublicKeyDecode(keyDer, &idx, &myKey, keySz);
+            if (ret == 0)
+                ret = wc_ecc_verify_hash(sig, sigSz, hash, hashSz, result, &myKey);
+            wc_ecc_free(&myKey);
+        }
+    }
+    return ret;
+}
+
+static int wolfIoT_ecc_shared_secret(WOLFSSL* ssl, struct ecc_key* otherKey,
+        unsigned char* pubKeyDer, word32* pubKeySz,
+        unsigned char* out, word32* outlen,
+        int side, void* ctx)
+{
+    int ret;
+    char *resp;
+    ecc_key   tmpKey;
+    IOTSAFE *iotsafe = &ssl->iotsafe;
+    byte keypair_slot = iotsafe->ecdh_keypair_slot;
+    byte pubkey_idx = iotsafe->peer_pubkey_slot;
+
+    (void)ctx;
+
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_shared_secret\n");
+
+    if (iotsafe->enabled) {
+        WOLFSSL_MSG("Generating ECDH key pair");
+        ret = iotsafe_gen_keypair(keypair_slot);
+        if (ret < 0) {
+            WOLFSSL_MSG("Error generating IoT-safe key pair");
+            return ret;
+        }
+
+        /* Importing generated public key */
+        ret = iotsafe_get_public_key(keypair_slot, &tmpKey);
+        if (ret < 0) {
+            return WC_HW_E;
+        }
+
+        /* Exporting generated public key into DER buffer */
+        if (wc_ecc_export_x963(&tmpKey, pubKeyDer, pubKeySz) < 0) {
+            WOLFSSL_MSG("Error exporting DER for key negotiation");
+            return -1;
+        }
+
+        /* Store received public key from other endpoint in applet */
+        ret = iotsafe_put_public_key(pubkey_idx, otherKey);
+        if (ret < 0) {
+            WOLFSSL_MSG("Error storing peer's public key");
+            return ret;
+        }
+        do {
+            ret = expect_ok("AT\r\n", 4);
+        } while (ret < 1);
+
+        /* Generating shared secret
+        */
+        iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_COMPUTE_DH, 0, 0);
+        iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PRIVKEY_ID, 1, &keypair_slot);
+        iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PUBKEY_ID, 1, &pubkey_idx);
+        iotsafe_cmd_complete(csim_cmd);
+        ret = expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp);
+        if (ret < 1) {
+            WOLFSSL_MSG("Unexpected reply in DH command");
+            return -1;
+        }
+        *outlen = hex_to_bytes(resp, out, ret / 2);
+        ret = 0;
+    } else {
+        ecc_key*  privKey = NULL;
+        ecc_key*  pubKey = NULL;
+        ret = wc_ecc_init(&tmpKey);
+        if (ret != 0) {
+            return ret;
+        }
+        /* for client: create and export public key */
+        if (side == WOLFSSL_CLIENT_END) {
+            WC_RNG rng;
```

Comment:
Lots of over-head setting up a new DRBG instance... Consider using the existing one in the WOLFSSL object with `WC_RNG* rng = wolfSSL_GetRNG(ssl)`.

Reply:
Good one, fixed as suggested everywhere the callbacks use a RNG. Worth mentioning that this is inspired from wolfssl/test.h default callbacks which don't use the session RNG.